### PR TITLE
Runtime + Compile + Macro Syntaxes for Variables and Parameters

### PR DESCRIPTION
**Pull request by thomhurst**
This provides strongly typed objects where users can call:

`Parameter.RuntimeExpression` => "parameters.Parameter"

`Parameter.CompileTimeExpression` => "{{ parameters.Parameter }}"

`Variable.MacroExpression` => "$(Variable)"

`Variable.RuntimeExpression` =>  "variables['Variable']"

`Variable.CompileTimeExpression` => "{{ variables.Variable }}

This gives the caller more control if they need it.

There's also new types with implicit conversions.

Strings implicitly convert to StringRuntimeExpression for string conditions.
StringRuntimeExpression is an IRuntimeExpression. This means conditions can accept these raw string expressions, or the new variable objects, or the new parameter objects. Then we can pull the correct run-time syntax out of these objects.

Variables now also serialize to Macro syntax so they can be used in scripts and such, but still serialize to runtime syntax for conditions and such.

This PR might seem quite convoluted but I couldn't think of a cleaner way to do this.
Let me know what you think.

See #232

### CODE REVIEW

**Review by premun — CHANGES_REQUESTED** [review]
I don't think this is convoluted at all. I think it's a really clever use of implicit conversions and I think this might be a priceless feature for Sharpliner!

However, I will need to do more testing to check for myself how it works with the 4 different combinations of the methods we need and whether the compiler can always figure things out.

In the meantime, I left some minor comments mostly around styling but there is one situation where this new unboxing does not work but I believe it can be fixed.

If you open `TemplateTests` there are some expressions in the very last test (`Conditioned_Parameters_Serialization_Test`). If you modify the pipeline used there and use the `parameters["container"]` strong typed notation, you will see the problem.
The problem is that we are creating there a `TemplateParameter` dictionary which is a dictionary that supports conditioned expressions. In this case, we are not implicitly casting into `string` because `TemplateParameter` expect an `object`. Then, the YAML serializer just serializes the into the `ParameterValue` object directly ending up with this:

```yaml
    parameterName: container
    runtimeExpression: parameters.container
    compileTimeExpression: ${{ parameters.container }}"
```

Now I _think_ you can solve this by making `ParameterValue` implement `IYamlConvertible` and implement the `Write` to serialize into `${{ parameters.container }}` if I'm not mistaken. But maybe it has another solution that I can't think of now.

Can you have a look at that?

File: src/Sharpliner/AzureDevOps/ConditionedExpressions/Interfaces/ICompileTimeExpression.cs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+namespace Sharpliner.AzureDevOps.ConditionedExpressions.Interfaces;
+
+public interface ICompileTimeExpression
```
Comment (premun) [inline src/Sharpliner/AzureDevOps/ConditionedExpressions/Interfaces/ICompileTimeExpression.cs:3]: Could you add a short XML doc to each of these and what they represent? It might be helpful for future

File: tests/Sharpliner.Tests/AzureDevOps/StrongTypedTests.cs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            }
+        };
+    }
+
+    [Fact]
+    public void Variable_And_Parameter_Serialize_Differently_Within_Scripts_And_Conditions_Test()
+    {
+        var yaml = new Pipeline_With_Strong_Variables_And_Parameters().Serialize();
+
+        yaml.Trim().Should().Be(@"parameters:
```
[the remark this answers is not in the document]
Comment (thomhurst) [reply tests/Sharpliner.Tests/AzureDevOps/StrongTypedTests.cs]: Done

File: src/Sharpliner/AzureDevOps/ConditionedExpressions/VariableValue.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    public string VariableName { get; }
+
+    internal VariableValue(string variableName)
+    {
+        VariableName = variableName;
+    }
+
+    public string RuntimeExpression => $"variables['{VariableName}']";
+
+    public string CompileTimeExpression => $"${{{{ variables.{VariableName} }}}}";
```
[the remark this answers is not in the document]
Comment (thomhurst) [reply src/Sharpliner/AzureDevOps/ConditionedExpressions/VariableValue.cs]: Done

File: src/Sharpliner/AzureDevOps/ConditionedExpressions/ParameterValue.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    public string ParameterName { get; }
+
+    internal ParameterValue(string parameterName)
+    {
+        ParameterName = parameterName;
+    }
+
+    public string RuntimeExpression => $"parameters.{ParameterName}";
+
+    public string CompileTimeExpression => $"${{{{ parameters.{ParameterName} }}}}";
```
[the remark this answers is not in the document]
Comment (thomhurst) [reply src/Sharpliner/AzureDevOps/ConditionedExpressions/ParameterValue.cs]: Done

File: src/Sharpliner/AzureDevOps/ConditionedExpressions/ParameterValue.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
+using Sharpliner.AzureDevOps.ConditionedExpressions.Interfaces;
+
+namespace Sharpliner.AzureDevOps.ConditionedExpressions;
+
+public class ParameterValue : IRuntimeExpression, ICompileTimeExpression
```
[the remark this answers is not in the document]
Comment (thomhurst) [reply src/Sharpliner/AzureDevOps/ConditionedExpressions/ParameterValue.cs]: Done

File: src/Sharpliner/AzureDevOps/ConditionedExpressions/Interfaces/ICompileTimeExpression.cs (hunk 1 above)
Comment (thomhurst) [reply src/Sharpliner/AzureDevOps/ConditionedExpressions/Interfaces/ICompileTimeExpression.cs:7]: Done

**Comment by thomhurst** [comment]
The IYamlConvertible worked !

File: src/Sharpliner/AzureDevOps/ConditionedExpressions/ConditionBuilder.cs [hunk 6]
```diff
@@ -129,13 +200,40 @@ public Condition<T> And(params string[] expressions)
...
+    public Condition<T> Contains(IRuntimeExpression needle, StringRuntimeExpression haystack)
+        => new ContainsCondition<T>(needle, haystack);
+
+    public Condition<T> Contains(StringRuntimeExpression needle, IRuntimeExpression haystack)
+        => new ContainsCondition<T>(needle, haystack);
+
+    public Condition<T> Contains(IRuntimeExpression needle, IRuntimeExpression haystack)
         => new ContainsCondition<T>(needle, haystack);
 
     public Condition In(string needle, string haystack)
```
Comment (premun) [inline src/Sharpliner/AzureDevOps/ConditionedExpressions/ConditionBuilder.cs:239]: So this is my fault, not yours, but we're missing some APIs here. Would you mind completing those too? We can also do it as part of another PR but maybe it makes sense?

Something like 
```csharp
    public Condition In(string needle, string haystack)
        => new InCondition(needle, haystack);

    public Condition In(string needle, params string[] haystack)
        => new InCondition(needle, haystack);

    public Condition NotIn(string needle, string haystack)
        => new NotInCondition(needle, haystack);

    public Condition NotIn(string needle, params string[] haystack)
        => new NotInCondition(needle, haystack);
```
(not sure these might not need changing to `IRuntimeExpression` then.

What should be possible is something like this:
```csharp
Variables =
{
    If.In(variables.Build.SourceBranch, "refs/heads/feature/", parameters["allowedTags"], variables["foo"], variables.Build.Reason)
        .Variable("feature", "on"),
}
```

File: src/Sharpliner/AzureDevOps/ConditionedExpressions/ConditionBuilder.cs [hunk 7]
```diff
@@ -147,10 +245,16 @@ public Condition NotIn(string needle, params string[] haystack)
...
+    public Condition<T> IsBranch(IRuntimeExpression branchName)
+        => Link(new BranchCondition<T>(branchName, true));
+
+    public Condition<T> IsNotBranch(StringRuntimeExpression branchName)
+        => Link(new BranchCondition<T>(branchName, false));
+
+    public Condition<T> IsNotBranch(IRuntimeExpression branchName)
         => Link(new BranchCondition<T>(branchName, false));
 
     public Condition<T> Greater(string first, string second)
```
Comment (premun) [inline src/Sharpliner/AzureDevOps/ConditionedExpressions/ConditionBuilder.cs:260]: `Greater` and `Less` should ideally also work with the new expressions

File: src/Sharpliner/AzureDevOps/ConditionedExpressions/ConditionBuilder.cs [hunk 8]
```diff
@@ -105,10 +158,28 @@ internal ConditionBuilder(Conditioned<T>? parent = null)
...
+    public Condition<T> NotEqual(IRuntimeExpression expression1, StringRuntimeExpression expression2)
+        => Link(new EqualityCondition<T>(false, expression1, expression2));
+
+    public Condition<T> NotEqual(StringRuntimeExpression expression1, IRuntimeExpression expression2)
+        => Link(new EqualityCondition<T>(false, expression1, expression2));
+
+    public Condition<T> NotEqual(IRuntimeExpression expression1, IRuntimeExpression expression2)
         => Link(new EqualityCondition<T>(false, expression1, expression2));
 
     public Condition<T> And(params Condition[] expressions)
```
Comment (premun) [inline src/Sharpliner/AzureDevOps/ConditionedExpressions/ConditionBuilder.cs:185]: `And` and `Or` in my opinion should not support the new stuff, so I'd keep them as-is

**Review by premun — COMMENTED** [review]
Looks good!

I found couple of more missing APIs but from my side and already missing in Sharpliner. Please let me know if you are rushing with having this PR in and we can merge right away and add the missing ones as a follow-up, or whether you could also add them as part of this PR.

File: src/Sharpliner/AzureDevOps/ConditionedExpressions/ConditionBuilder.cs (hunk 8 above)
Comment (thomhurst) [reply src/Sharpliner/AzureDevOps/ConditionedExpressions/ConditionBuilder.cs:242]: Done

File: src/Sharpliner/AzureDevOps/ConditionedExpressions/ConditionBuilder.cs (hunk 7 above)
Comment (thomhurst) [reply src/Sharpliner/AzureDevOps/ConditionedExpressions/ConditionBuilder.cs]: Done

File: src/Sharpliner/AzureDevOps/ConditionedExpressions/ConditionBuilder.cs (hunk 6 above)
Comment (thomhurst) [reply src/Sharpliner/AzureDevOps/ConditionedExpressions/ConditionBuilder.cs]: I had to add two extra methods to accept a params object[] array. This is so the caller can mix variables/params (IRuntimeExpression) and strings within the same array.

But this should be done now if you'd like to take a look. Added a test for your above scenario too.

**Comment by premun** [comment]
@thomhurst this is really amazing, thank you!

I have actually heard many times online from people that mixing these types of variables causes them a lot of headache so I will update the README later to list this as one of the main features.

**Review by premun — APPROVED** [review]
